src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -14,211 +24,3020 @@ namespace Microsoft.CodeAnalysis.CSharp
+        /// that label with the state when we fall into the label.  Entries are created when the
+        /// label is encountered.  One case deserves special attention: when the destination of the
+        /// branch is a label earlier in the code, it is possible (though rarely occurs in practice)
+        /// that we are changing the state at a label that we've already analyzed. In that case we
+        /// run another pass of the analysis to allow those changes to propagate. This repeats until
+        /// no further changes to the state of these labels occurs.  This can result in quadratic
+        /// performance in unlikely but possible code such as this: "int x; if (cond) goto l1; x =
+        /// 3; l5: print x; l4: goto l5; l3: goto l4; l2: goto l3; l1: goto l2;"
+        /// </summary>
+        private readonly PooledDictionary<LabelSymbol, TLocalState> _labels;
+
+        /// <summary>
+        /// Set to true after an analysis scan if the analysis was incomplete due to state changing
+        /// after it was used by another analysis component.  In this case the caller scans again (until
+        /// this is false). Since the analysis proceeds by monotonically changing the state computed
+        /// at each label, this must terminate.
+        /// </summary>
+        protected bool stateChangedAfterUse;
+
+        /// <summary>
+        /// All of the labels seen so far in this forward scan of the body
+        /// </summary>
+        private PooledHashSet<BoundStatement> _labelsSeen;
+
+        /// <summary>
+        /// If we are tracking exceptions, then by convention the first entry in the pending branches
+        /// buffer contains a summary of the states that can arise from exceptions.
+        /// </summary>
+        private readonly bool _trackExceptions;
+
+        /// <summary>
+        /// Pending escapes generated in the current scope (or more deeply nested scopes). When jump
+        /// statements (goto, break, continue, return) are processed, they are placed in the
+        /// pendingBranches buffer to be processed later by the code handling the destination
+        /// statement. As a special case, the processing of try-finally statements might modify the
+        /// contents of the pendingBranches buffer to take into account the behavior of
+        /// "intervening" finally clauses.
+        /// </summary>
+        protected ArrayBuilder<PendingBranch> PendingBranches { get; private set; }
+
+        /// <summary>
+        /// The definite assignment and/or reachability state at the point currently being analyzed.
+        /// </summary>
+        protected TLocalState State;
+        protected TLocalState StateWhenTrue;
+        protected TLocalState StateWhenFalse;
+        protected bool IsConditionalState;
+
+        private readonly bool _nonMonotonicTransfer; // for the data flows out walker, we track unassignments as well as assignments
```

Comment:
some fields seem to be doc'ed as being sepcific to a particular walker.  is there a reason we don't just put those fields in those particular walkers?

Reply:
This is common to a few walkers.  This comment should clarify what it means: the transfer function is *not* monotonic, in the sense that it can change the state in either direction in the lattice.  If the transfer function is monotonic, the transfer function can only change the state toward the bottom (e.g. by transferring control, leaving the state "unreachable", or by assigning a variable, leaving it definitely assigned). Reachability and definite assignment are monotonic, and permit a more efficient analysis.  Region analysis and nullable analysis are not monotonic.  This is just an optimization; we could treat all of them as nonmonotonic without much loss of performance.  In fact, this only affects the analysis of (relatively rare) try statements, and is only a slight optimization.  We might consider eliminating it.

---
In reply to: [236562382](https://github.com/dotnet/roslyn/pull/31389#discussion_r236562382) [](ancestors = 236562382)

Reply:
Most of Neal's comment would be worthwhile to include in the comment for `_nonMonotonicTransfer`.

---
In reply to: [236813973](https://github.com/dotnet/roslyn/pull/31389#discussion_r236813973) [](ancestors = 236813973,236562382)